uportal-war/src/main/java/org/jasig/portal/groups/pags/testers/AdHocGroupTester.java

```diff
@@ -0,0 +1,172 @@
+import org.jasig.portal.security.IPerson;
+import org.jasig.portal.services.GroupService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentSkipListSet;
+
+/**
+ * Immutable PAGS Tester for inclusive/exclusive membership in sets of groups. To avoid infinite recursion,
+ * calls to {@code test()} are tracked by parameters plus thread ID.
+ *
+ * @author  Benito J. Gonzalez <bgonzalez@unicon.net>
+ * @see     org.jasig.portal.groups.pags.IPersonTester
+ * @see     org.jasig.portal.groups.pags.dao.EntityPersonAttributesGroupStore
+ * @see     org.jasig.portal.groups.pags.dao.IPersonAttributesGroupTestGroupDefinition
+ * @since   4.3
+ */
+public final class AdHocGroupTester implements IPersonTester {
+
+    private static final Logger logger = LoggerFactory.getLogger(AdHocGroupTester.class);
+    private static final Set<String> currentPersons = new ConcurrentSkipListSet<>();
+    private final Set<IEntityGroup> includes = new HashSet<>();
+    private final Set<IEntityGroup> excludes = new HashSet<>();
+    private final String hashcode;
+
+    public AdHocGroupTester(IPersonAttributesGroupTestDefinition definition) {
+        Validate.notNull(definition.getIncludes());
+        Validate.notNull(definition.getExcludes());
+        collectGroups(definition.getIncludes(), this.includes, true);
+        collectGroups(definition.getExcludes(), this.excludes, false);
+        hashcode = calcHashcode();
+    }
+
+    /**
+     * Given a set of group names, find the entity groups and add them to the {@code groups} collection.
+     * If a group is not found, either log a warning or throw an {@code IllegalArgumentException} based on
+     * the {@code throwOnFail} parameter.
+     *
+     * @param groupNames    Set of group names to find
+     * @param groups        Set to add named groups from {@link GroupService}
+     * @param throwOnFail   flag to indicate whether to log or throw exception if a group is not found
+     */
+    private void collectGroups(Set<String> groupNames, Set<IEntityGroup> groups, boolean throwOnFail) {
+        for (String groupName : groupNames) {
+            IEntityGroup entityGroup = findGroupByName(groupName);
+            if (entityGroup != null) {
+                groups.add(entityGroup);
+            } else {
+                if (throwOnFail) {
+                    logger.error("Could not find group named {}", groupName);
+                    throw new IllegalArgumentException("Could not find group named " + groupName);
+                } else {
+                    logger.warn("Could not find group named {}", groupName);
+                }
+            }
+        }
+    }
+
+    /**
+     * Create a hashcode based on the includes/excludes groups.
+     *
+     * @return      hashcode for this instance
+     */
+    private String calcHashcode() {
```

Comment:
I will change this method to something like "calcGroupsHash".

Testing membership person membership will create several instances of testers that represent the same test.  When the thread recurses to the same AdHocGroupTest condition, it will create a second object that is different than others in the call stack. 

Once the PAGS group store is refactored to reuse testers, this would work.